src/include/function/cast/functions/cast_functions.h

```diff
@@ -0,0 +1,247 @@
+#pragma once
+
+#include <cassert>
+
+#include "cast_string_non_nested_functions.h"
+#include "common/exception/conversion.h"
+#include "common/exception/overflow.h"
+#include "common/string_format.h"
+#include "common/type_utils.h"
+#include "common/types/blob.h"
+#include "common/vector/value_vector.h"
+
+namespace kuzu {
+namespace function {
+
+struct CastToString {
+    template<typename T>
+    static inline std::string castToString(T& input, const common::ValueVector& inputVector) {
+        return common::TypeUtils::toString(input, (void*)&inputVector);
+    }
+
+    template<typename T>
+    static inline void operation(T& input, common::ku_string_t& result,
+        common::ValueVector& inputVector, common::ValueVector& resultVector) {
+        std::string resultStr = castToString(input, inputVector);
+        if (resultStr.length() > common::ku_string_t::SHORT_STR_LENGTH) {
+            result.overflowPtr = reinterpret_cast<uint64_t>(
+                common::StringVector::getInMemOverflowBuffer(&resultVector)
+                    ->allocateSpace(resultStr.length()));
+        }
+        result.set(resultStr);
+    }
+};
+
+struct CastDateToTimestamp {
+    static inline void operation(common::date_t& input, common::timestamp_t& result) {
+        result = common::Timestamp::fromDateTime(input, common::dtime_t{});
+    }
+};
+
+struct CastToDouble {
+    template<typename T>
+    static inline void operation(T& input, double_t& result) {
+        if (!tryCastWithOverflowCheck(input, result)) {
+            throw common::OverflowException{common::stringFormat(
+                "Value {} is not within DOUBLE range", common::TypeUtils::toString(input).c_str())};
+        }
+    }
+};
+
+template<>
+inline void CastToDouble::operation(common::int128_t& input, double_t& result) {
+    if (!common::Int128_t::tryCast(input, result)) { // LCOV_EXCL_START
+        throw common::OverflowException{common::stringFormat(
+            "Value {} is not within DOUBLE range", common::TypeUtils::toString(input).c_str())};
```

Comment:
why is this excluded from coverage?

Reply:
int128 to double can't trigger overflow since int128 is always in double range

Reply:
`INT128 is always in double range` is not true.
Large integer numbers in double are not consecutive because of the precision. For example: 2^100 may be representable in double but 2^100 + 1 cannot be represented in double. I am not sure what is the expected behaviour in  this case.

Reply:
Converting to double and losing precision is not surprising. Postgres[1] says that the number stored may be inexact and to use numeric for exactness.

Let's add a comment for why this is unreachable.

    [1]: https://www.postgresql.org/docs/current/datatype-numeric.html

Reply:
duckdb always return true and so for us

Reply:
│ CAST(CAST('170141183460469231731687303715884105727' AS HUGEINT) AS DOUBLE) │
|:---:|
│                                   double                               |
│                                                     1.7014118346046923e+38 |

Reply:
The point Ziyi is making is that if you add 1 to the integer there, the actual value stored will likely be the same because doubles are not arbitrary precision.